src/coreclr/jit/morphblock.cpp

```diff
@@ -0,0 +1,1551 @@
+    }
+#endif // TARGET_ARM
+
+    // Don't use field by field assignment if the src is a call,
+    // lowering will handle it without spilling the call result into memory
+    // to access the individual fields.
+    //
+    if (m_src->OperGet() == GT_CALL)
+    {
+        JITDUMP(" src is a call");
+        requiresCopyBlock = true;
+    }
+
+    // If we passed the above checks, then we will check these two
+    if (!requiresCopyBlock)
+    {
+        // It is not always profitable to do field by field init for structs that are allocated to memory.
+        // A struct with 8 bool fields will require 8 moves instead of one if we do this transformation.
+        // A simple heuristic when field by field copy is prefered:
+        // - if fields can be enregistered;
+        // - if the struct has GCPtrs (block copy would be done via helper that is expensive);
+        // - if the struct has only one field.
+        bool dstFldIsProfitable =
+            ((m_dstVarDsc != nullptr) &&
+             (!m_dstVarDsc->lvDoNotEnregister || m_dstVarDsc->HasGCPtr() || (m_dstVarDsc->lvFieldCnt == 1)));
+        bool srcFldIsProfitable =
+            ((m_srcVarDsc != nullptr) &&
+             (!m_srcVarDsc->lvDoNotEnregister || m_srcVarDsc->HasGCPtr() || (m_srcVarDsc->lvFieldCnt == 1)));
+        // Are both dest and src promoted structs?
+        if (m_dstDoFldAsg && m_srcDoFldAsg && (dstFldIsProfitable || srcFldIsProfitable))
+        {
+            // Both structs should be of the same type, or have the same number of fields of the same type.
+            // If not we will use a copy block.
+            bool misMatchedTypes = false;
+            if (m_dstVarDsc->GetStructHnd() != m_srcVarDsc->GetStructHnd())
+            {
+                if (m_dstVarDsc->lvFieldCnt != m_srcVarDsc->lvFieldCnt)
+                {
+                    misMatchedTypes = true;
+                }
+                else
+                {
+                    for (int i = 0; i < m_dstVarDsc->lvFieldCnt; i++)
+                    {
+                        LclVarDsc* destFieldVarDsc = m_comp->lvaGetDesc(m_dstVarDsc->lvFieldLclStart + i);
+                        LclVarDsc* srcFieldVarDsc  = m_comp->lvaGetDesc(m_srcVarDsc->lvFieldLclStart + i);
+                        if ((destFieldVarDsc->lvType != srcFieldVarDsc->lvType) ||
+                            (destFieldVarDsc->lvFldOffset != srcFieldVarDsc->lvFldOffset))
+                        {
+                            misMatchedTypes = true;
+                            break;
+                        }
+                    }
+                }
+                if (misMatchedTypes)
+                {
+                    requiresCopyBlock = true; // Mismatched types, leave as a CopyBlock
+                    JITDUMP(" with mismatched types");
+                }
+            }
+        }
+        else if (m_dstDoFldAsg && dstFldIsProfitable)
+        {
+            // Match the following kinds of trees:
```

Comment:
Just a note: would be nice to update these to the new dump format someday (which is much better in my opinion).

Reply:
Thanks, I will change this method further and include it. For now, I need to do the biggest code movements, which we will merge mostly based on the fact that there are no diffs rather than checking each small code piece that was moved.
After that these methods would become available for more changes.